### PR TITLE
Fix buffer overflow when reading EXIF metadata

### LINKED ISSUE #550: Buffer overflow when reading some EXIF metadata
I stumbled upon some images that trigger a server internal error when cantaloupe tries to read their EXIF metadata.

Here is the stacktrace

```
java.lang.IndexOutOfBoundsException: off < 0 || len < 0 || off+len > b.length || off+len < 0!
	at java.desktop/javax.imageio.stream.MemoryCacheImageInputStream.read(MemoryCacheImageInputStream.java:100)
	at edu.illinois.library.cantaloupe.image.exif.Reader.readBytes(Reader.java:159)
	at edu.illinois.library.cantaloupe.image.exif.Reader.read(Reader.java:145)
	at edu.illinois.library.cantaloupe.image.exif.Reader.read(Reader.java:97)
	at edu.illinois.library.cantaloupe.processor.TurboJpegProcessor.readMetadata(TurboJpegProcessor.java:275)
	at edu.illinois.library.cantaloupe.processor.TurboJpegProcessor.readInfo(TurboJpegProcessor.java:263)
	at edu.illinois.library.cantaloupe.cache.InfoService.readInfo(InfoService.java:220)
	at edu.illinois.library.cantaloupe.cache.InfoService.getOrReadInfo(InfoService.java:153)
	at edu.illinois.library.cantaloupe.cache.CacheFacade.getOrReadInfo(CacheFacade.java:67)
	at edu.illinois.library.cantaloupe.resource.AbstractRequestHandler.getOrReadInfo(AbstractRequestHandler.java:39)
	at edu.illinois.library.cantaloupe.resource.InformationRequestHandler.handle(InformationRequestHandler.java:287)
	at edu.illinois.library.cantaloupe.resource.iiif.v2.InformationResource.doGET(InformationResource.java:85)
	at edu.illinois.library.cantaloupe.resource.HandlerServlet.handle(HandlerServlet.java:97)
	at edu.illinois.library.cantaloupe.resource.HandlerServlet.doGet(HandlerServlet.java:35)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:791)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:773)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:905)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### PR DESCRIPTION
**Pull request by La0**
Closes #550

### CODE REVIEW

**Review by janhoy — APPROVED** [review]
Looks good

**Comment by keskadaniel** [comment]
@janhoy when do you plan to publish new release? Will you include this fix? 
I have the same error and it affects a lot of pictures in my app.

**Comment by janhoy** [comment]
I'm not affiliated with the project, and sadly it looks like University of Illinois has abandoned the project and are also not willing to entrust external developers write access to the code to ressurect it and turn it into a true OSS community. The wish for tight control of their "baby" thus leads to the baby's death. Not uncommon but still sad to see.

**Comment by La0** [comment]
Would you be interested in a fork ?

We are thinking about doing this at my company, as we already maintain some patches, plugins and a docker image for cantaloupe...

**Comment by glenrobson** [comment]
Hi Both,

I've been working on a security update which is currently building here (comments / testing welcome!): 

https://github.com/glenrobson/cantaloupe/pull/1

I've contacted Illinois to see if we can get this released. Once this security update is out hopefully we can then turn to merging in the pull requests from the community. 

Ill keep you updated.

Thanks

Glen

File: src/main/java/edu/illinois/library/cantaloupe/image/exif/Reader.java [hunk 1]
```diff
@@ -157,6 +157,10 @@ private byte[] readBytes(int length) throws IOException {
         byte[] data = new byte[length];
         int n, offset = 0;
         while ((n = inputStream.read(data, offset, data.length - offset)) < offset) {
+            if (n == -1) {
+                LOGGER.trace("readBytes(): No more data to read");
+                break;
+            }
             offset += n;
         }
```
Comment (adam-vessey) [inline src/main/java/edu/illinois/library/cantaloupe/image/exif/Reader.java:165]: Have recently stumbled across a similar implementation in [4.1's `JPEG2000MetadataReader`](https://github.com/cantaloupe-project/cantaloupe/blob/release/4.1/src/main/java/edu/illinois/library/cantaloupe/processor/codec/JPEG2000MetadataReader.java#L409-L417) which was breaking due to _not_ reading the full structure. We're presently moving to patch it, with something more like:

```suggestion
        while ((n = inputStream.read(data, offset, data.length - offset)) > -1) {
            offset += n;
            if (offset >= length) {
              break;
            }
        }
        if (offset < length) {
          throw new IOException("Failed to read the expected amount of data.");
        }
```

---

Not really a Java dev, but looks like there's a `inputStream.readFully()` that might be expected to deal with fully reading the content, and deal with the boundary conditions?

**Comment by keskadaniel** [comment]
Hi @glenrobson,
I saw new release and I need to ask when do you plan to merge pull requests?
Thank you for doing that!
Daniel

**Comment by glenrobson** [comment]
Hi Daniel, I'm in contact with the Illinois developers. Our intial agreement was to do a security update release but I'm hoping we can now turn to fixes. Ill keep you updated.

Thanks

Glen

**Comment by glenrobson** [comment]
Hi @La0, now we are in a position to start taking in pull requests. Could you review @adam-vessey's suggestion. Also would it be possible to add a test to catch this exception?

Thanks Glen

File: src/main/java/edu/illinois/library/cantaloupe/image/exif/Reader.java (hunk 1 above)
Comment (adam-vessey) [reply src/main/java/edu/illinois/library/cantaloupe/image/exif/Reader.java:165]: @glenrobson: I've thrown together https://github.com/cantaloupe-project/cantaloupe/pull/659 making use of `inputStream.readFully()` here, and a few other places, which I think should do the trick for this and a couple of other places.

**Comment by La0** [comment]
Sure, I'll revisit this issue next week, i'll keep you updated here

**Comment by La0** [comment]
I'm closing this in favor of #659 which is more exhaustive

**Pull request closed without merging**